rclcpp/include/rclcpp/subscription.hpp

```diff
@@ -325,8 +326,31 @@ class Subscription : public SubscriptionBase
     const std::shared_ptr<rclcpp::SerializedMessage> & serialized_message,
     const rclcpp::MessageInfo & message_info) override
   {
-    // TODO(wjwwood): enable topic statistics for serialized messages
+    if (matches_any_intra_process_publishers(&message_info.get_rmw_message_info().publisher_gid)) {
+      // In this case, the message will be delivered via intra process and
+      // we should ignore this copy of the message.
+      return;
+    }
+
+    std::chrono::time_point<std::chrono::system_clock> now;
+    if (subscription_topic_statistics_) {
+      // get current time before executing callback to
+      // exclude callback duration from topic statistics result.
+      now = std::chrono::system_clock::now();
+    }
+
     any_callback_.dispatch(serialized_message, message_info);
+
+    ROSMessageType deserialized_message;
+    auto serializer = rclcpp::Serialization<ROSMessageType>();
+    serializer.deserialize_message(serialized_message.get(), &deserialized_message);
```

Comment:
I know this is unavoidable given the current APIs, but this is going to make serialized messages vastly more expensive.  Consider a large, multi-megabyte pointcloud; we are deserializing this for every single message now.  I don't think that this is a viable strategy, as one of the goals of serialized messages is to avoid exactly that kind of deserialization (which is what we do in the normal, non-serialized message path).

Reply:
@clalancette From one hand, yes it will make getting serialized messages out of a regular strictly typed subscription more expensive due to the deserialization. But on another hand, think about in what exact cases?
When all of the following conditions are true
1. Topics statistics enabled!
2. The user getting serialized messages out of the strictly typed subscription which is odd and very unusual. In ROS 2 core packages we are not using such a scenario AFAIK and I can't imagine from the top of my head why it would be needed for the user. For reference in rosbag2 we are using `GenericSubscription` which is not affected by these changes.

Reply:
> The user getting serialized messages out of the strictly typed subscription which is odd and very unusual.

If rosbag2 implements using `GenericSubscription` and this really is as uncommon as you say, then maybe this should be down in the `handle_serialized_message` method in that [class](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/generic_subscription.cpp#L50-L55). We most likely don't want to be writing features for very explicit scenarios when we can cover most sections with a valid general solution.

That being said, while I won't have to include any of the intraprocess stuff, I am interested in passing these serialized messages to topic statistics, while staying serialized. Would we do something similar to the passing chain of `SerializedMessage` to the [`dispatch` call](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/include/rclcpp/any_subscription_callback.hpp) in any_subscription_callback, except we would do it with [`SubscriptionTopicStatistics`](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/include/rclcpp/topic_statistics/subscription_topic_statistics.hpp#L107-L115) and `handle_message`. Something along the lines of : 
```
virutal void handle_message(
    const rclcpp::SerializedMessage & received_message,
    const rclcpp::Time now_nanoseconds) const
  {
    std::lock_guard<std::mutex> lock(mutex_);
    for (const auto & collector : subscriber_statistics_collectors_) {
      collector->OnMessageReceived(received_message, now_nanoseconds.nanoseconds());
    }
  }
```
From there we would also have to implement a `SubscriptionTopicStatistics` object within the `GenericSubscription` class with a `typename T` of rclcpp::Serialized message, assuming it would still work with it.

Reply:
@CursedRock17 As regards to the 
> If rosbag2 implements using GenericSubscription and this really is as uncommon as you say, then maybe this should be down in the handle_serialized_message method in that [class](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/generic_subscription.cpp#L50-L55).

In this case, I would **not** approve this motion since in the current design and implementation it will really affect recording performance when libstatistics is enabled.

As regards your request and proposal 
> I am interested in passing these serialized messages to topic statistics, while staying serialized. 

And what are lib topic statistics supposed to do with those serialized messages?   
Deserialize inside? - IMO it will be the same performance burden but on the downstream layer and it doesn't make sense at all.

As I have said above if will extend or change API for the `libstatistic_collector` it will make sense to pass `message_info` structure instead of the real message.

Reply:
using `message_info` does help here. and i think that is the way it should be? enabling statistics requires deserialization for message packet will be huge burden for user application. system should honor the behavior requested by the application to keep the message payload serialized.

Reply:
So is this something we bring to the [`libstatistic_collector`](www.github.com/ros-tooling/libstatistics_collector) library now, or do we wait for for necessary changes to be made? The library is quite low on issues and pull requests as of now, but I wouldn't want to release a bunch of breaking changes just for a more explicit addition to rclcpp.

Reply:
@CursedRock17 We have a lack of resources for adding a new feature in `libstatistics_collector` if you have time and are willing to help it will be very appreciated.  
I think the right way to handle it will be to create a new issue and PR in `libstatistic_collector` repo for adding a new API with `message_info` as a parameter and then deprecate the old one.

Reply:
I would definitely be willing to work on an issue and pull request for the library, but where else could `libstatistic_collector` get us out of this feature, for example could we then use it for the other `handle_x_message` functions after we deprecate the old `const T & recieved_message`?

Reply:
As regards
> could we then use it for the other handle_x_message functions after we deprecate the old const T & recieved_message

I am a bit confused with the question. Why not? The `message_info` seems available in all handlers.

Reply:
Oh I knew we could use `message_info` just wanted some clarification of proceeding that way, it shouldn't change the time/space complexity of those functions either, because we're passing both `message_info` and `received_message` by const ref. I'm going to go ahead and add both an issue and pr to the library.

Reply:
according to this discussion, what about using `rmw_message_info_t` for statistics collector? that said, we can remove the dependency on `message` from here (agnostic from `CallbackMessageT`). that seems to be reasonable change to me, and clear design.

Reply:
Just for clarification, do you mean the TopicStatistic class within the rclcpp or the TopicStatistic in `libcollector` library. I brought up the type utilized within the issue I opened for that library. I'm not entirely familiar with the rmw_message_info object, could we just make changes in this library without having to change the `libcollector` library's method `OnMessageReceived`, because if so we should just make that change here in rclcpp.

Reply:
> do you mean the TopicStatistic class within the rclcpp or the TopicStatistic in libcollector library. 

`libstatisticscollector`.

> could we just make changes in this library without having to change the libcollector library's method OnMessageReceived

libcollector library's method OnMessageReceived is the one which takes `CallbackMessageT`. that is the reason that we need to deserialize the message here. probably we need to change that anyway to take `rmw_message_info_t` like structure. (i am saying this since libcollector library is not dependent on rmw, should not be, i think)

Reply:
Perfect, yea I agree with what you're saying. I linked the issue I made in `libcollector`, just to make sure it's maintainers were ok with us changing to `rmw_message_info`. Also like you said `CallbackMessageT`and all the templating for those function should be deprecated since we'll already know the type. So once they agree in the issue and I make the PR, I'll come back here and change our `handle_x_message` functions.